text/2022-07-25-PROPOSAL-build-glyphs-in-rust.md

```diff
@@ -0,0 +1,128 @@
+
+For maximum impact, do it horizontally, impacting as close as possible to all Google Fonts projects without having to touch them one by one to alter their bespoke build processes. If we must touch them, we should move them to a consistent build structure.
+
+Bias to heavy lifting done in Rust as part of Oxidize. A layer of Python orchestration is acceptable.
+
+Bias to incremental progress, no big bang integration.
+Writing a new compiler in isolation while everyone continues to use fontmake, and then abruptly switching projects over
+(big bang integration) delays impact and increases risk.
+An approach that incrementally improves fontmake has significant appeal.
+
+## Proposal
+
+Migrate the core for each glyph loop
+([varLib._add_gvar](https://github.com/fonttools/fonttools/blob/455158f2bfd9eae2135a5f85bb96549a545cac82/Lib/fontTools/varLib/__init__.py#L222))
+in fontmake from Python to Rust.
+
+* Expose Rust to Python.
+* Let Rust handle the parallelism.
+* Build directly from UFO sources, optionally passing through some intermediary representation.
+* Do NOT use the static TTFs at master locations to build glyph tables.
+  * the static TTFs can still exist, but their glyph work should be brought as close as possible to zero (empty glyphs?)
+
+The next step is to get approval from the fontmake owners to make such changes. @anthrotype basically :D
+
+## Context
+
+### How do you make a variable font anyway
+
+Complicating this, the core approach we take to build a variable font is excitingly indirect (https://simoncozens.github.io/compiling-variable-fonts/):
+
+   1. Generate static TTFs at master locations
+   1. Merge those TTFs to create a final VF
+
+As far as the author can tell there is no requirement for the statics to ever exist, we do it this way purely because
+it was a relatively simple way to get fontmake to support VF. In the end we should go from sources → final VF without this step.
+
+A concrete example may help, consider the N in [Josefin Slab](https://fonts.google.com/specimen/Josefin+Slab)
+(https://github.com/TypeNetwork/Josefinslab) upright:
+
+* [sources/JosefinSlab-Thin.ufo/glyphs/N_.glif](https://github.com/davelab6/josefinslab/blob/master/sources/JosefinSlab-Thin.ufo/glyphs/N_.glif) defines the curve at thin (N)
+* [sources/JosefinSlab-Bold.ufo/glyphs/N_.glif](https://github.com/davelab6/josefinslab/blob/master/sources/JosefinSlab-Bold.ufo/glyphs/N_.glif) defines the curve at bold (N)
+* [sources/JosefinSlab.designspace](https://github.com/davelab6/josefinslab/blob/master/sources/JosefinSlab.designspace) defines how the parts fit together
+   * JosefinSlab-Thin.ufo is at weight 14
+   * JosefinSlab-Bold.ufo is at weight 83
+   * Font weights are mapped onto the space defined by the UFO
+      * Default 100 maps to 14
+      * 300 maps to 28, 400 to 41, etc
+         * These aren't where you'd land by lerp, we need to watch how we set up interpolation
```

Comment:
Avar1 table might be relevant

Reply:
@anthrotype if *all* we do is migrate the glyph loop what are the implications of this mapping? For step one can this be done via avar with that code left untouched in Python-land?

Reply:
To produce the gvar deltas, a VariationModel needs to be created from a list of axes and master locations along those. These locations need to be in _normalized_ coordinates (-1.0 for min, 0.0 for default, +1.0 for max).

The designspace can contain optional axes `<map>` elements that are used to create an avar1 table; as you know, this "warps" the default normalization from user-scale coordinates (as found in the fvar table) to normalized coordinates, allowing to add intermediate maps in between min/default/max (the latter are fixed by definition) that "aren't where you'd land by lerp", to obtain a piecewise lerp basically.

There is a third "space" that we sometimes call internal, or design-space, which are the arbitrary numbers used to define the locate the master sources in the designspace; they can be anything a designer pleases and are not tied neither to a registered axis's official scale (e.g. 'normal' wdth must be 100 in user coordinates, as in percents, but can be any number in design coordinates), nor to the -1.0,0.0,+1.0 normalized scale.
In order to normalize design-space locations and map them to the -1.0,0.0,+1.0 range used internally by variation tables, we simply take the min, default and maximum design coordinate of each axis, map them respectively to -1.0, 0.0 and +1.0, then do a straightfoward lerp for the in-between design-space coordinates.

This is the fonttools varLib code that normalizes the master locations:
https://github.com/fonttools/fonttools/blob/455158f2bfd9eae2135a5f85bb96549a545cac82/Lib/fontTools/varLib/__init__.py#L801-L814

So to answer your question, the avar mappings are _not_ relevant to get normalized master locations, but only when we want to normalize the _user_-space coordinates (like the fvar's axis min/default/max, or named instances locations). That's assuming we already have internal, design locations for the masters, which is the case in the designspace format.

> For step one can this be done via avar with that code left untouched in Python-land?
 
yes

Reply:
Excellent clarification, thank you:)